### PR TITLE
Added ExchangeCodeAsync Method.

### LINKED ISSUE #529: Untappd not working
AspNet.Security.OAuth.Untappd is not working.
Reason seems to be that ExchangeCodeAsync needs to work slightly different.
The request needs to be HttpMethod.Get instead of Post and the parameters need to be transmitted as a QueryString according to https://untappd.com/api/docs#authentication

I add this issue and have a PR ready where I will link to this issue.

### PR DESCRIPTION
**Pull request by PatGet**
Added ExchangeCodeAsync Method that uses HttpMethod.Get and QueryString according to https://untappd.com/api/docs#authentication

Fix for the JSON structure within CreateTicketAsync. The Root Element is "response".
Changed bundle.json for the tests.

Issue: https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/issues/529

### CODE REVIEW

File: src/AspNet.Security.OAuth.Untappd/UntappdAuthenticationHandler.cs
Comment (LeaFrock) [inline src/AspNet.Security.OAuth.Untappd/UntappdAuthenticationHandler.cs]: I think it's not a good habit to write like this,  even if this method only handles the error situation, since the underlying library should pay as much attention to performance as possible. 

```C#

string rspText = await response.Content.ReadAsStringAsync();
var output = new StringBuilder(64) // give a suitable capacity
    .Append("Status: ")
    .Append(response.StatusCode) // use specific overload methods
    .Append(';') // use Append(char c)
    .....

```

File: src/AspNet.Security.OAuth.Untappd/UntappdAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.Untappd/UntappdAuthenticationHandler.cs]: Newer initializer syntax is preferred.

```suggestion
                ["client_id"] = Options.ClientId,
                ["redirect_uri"] = context.RedirectUri,
                ["client_secret"] = Options.ClientSecret,
                ["code"] = context.Code,
                ["grant_type"] = "authorization_code",
```

File: src/AspNet.Security.OAuth.Untappd/UntappdAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.Untappd/UntappdAuthenticationHandler.cs]:
```suggestion
            using var requestContent = new FormUrlEncodedContent(tokenRequestParameters!);
```

File: src/AspNet.Security.OAuth.Untappd/UntappdAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.Untappd/UntappdAuthenticationHandler.cs]: Use `QueryHelpers.AddQueryString()` instead. Example:

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/24743fb66bdb350c424e7d7f5e8ed67b3f4a60a1/src/AspNet.Security.OAuth.Weixin/WeixinAuthenticationHandler.cs#L62-L66

File: src/AspNet.Security.OAuth.Untappd/UntappdAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.Untappd/UntappdAuthenticationHandler.cs]:
```suggestion
            using var response = await Backchannel.SendAsync(requestMessage, Context.RequestAborted);
```

File: src/AspNet.Security.OAuth.Untappd/UntappdAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.Untappd/UntappdAuthenticationHandler.cs]: Please use the same failure handling pattern as our other providers. Example:

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/9e1fef49de1abd808d2d6cfde26a1c66ce51c6da/src/AspNet.Security.OAuth.GitHub/GitHubAuthenticationHandler.cs#L86-L95

File: src/AspNet.Security.OAuth.Untappd/UntappdAuthenticationHandler.cs
Comment (martincostello) [reply src/AspNet.Security.OAuth.Untappd/UntappdAuthenticationHandler.cs]: If we were to keep this specific approach I'd agree - this isn't using the full benefits of `StringBuilder` the way you've written it. However, this method should be removed and the error handling should follow the approach used elsewhere for consistency.

File: test/AspNet.Security.OAuth.Providers.Tests/Untappd/bundle.json
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/Untappd/bundle.json:40]: Nit: please fix the indentation.

**Review by martincostello — CHANGES_REQUESTED** [review]

File: src/AspNet.Security.OAuth.Untappd/UntappdAuthenticationHandler.cs
Comment (PatGet) [reply src/AspNet.Security.OAuth.Untappd/UntappdAuthenticationHandler.cs]: Go tell the the ASP.Net Core Team, i copied both Methods from there ;-)
https://github.com/dotnet/aspnetcore/blob/a24aed7e5459acc1417a0e490a80e0472d498b2b/src/Security/Authentication/OAuth/src/OAuthHandler.cs

Comment (PatGet) [reply src/AspNet.Security.OAuth.Untappd/UntappdAuthenticationHandler.cs]: Going to replace Error handling as @martincostello suggested to keep it similar.

File: src/AspNet.Security.OAuth.Untappd/UntappdAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.Untappd/UntappdAuthenticationHandler.cs]:
```suggestion
                                  /* Status: */ response.StatusCode,
                                  /* Headers: */ response.Headers.ToString(),
                                  /* Body: */ await response.Content.ReadAsStringAsync(Context.RequestAborted));
```

File: test/AspNet.Security.OAuth.Providers.Tests/Untappd/bundle.json
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/Untappd/bundle.json]:
```suggestion
        }
```

**Review by martincostello — APPROVED** [review]

**Comment by martincostello** [comment]
Thanks for the fix @PatGet 👍

Once this is merged could you just give it a quick check with your application using a build from [our MyGet feed](https://www.myget.org/feed/aspnet-contrib/package/nuget/AspNet.Security.OAuth.Untappd) with the fix? If all is OK I'll publish a 5.0.3 release with the fix to NuGet.org tomorrow.

**Comment by PatGet** [comment]
> Thanks for the fix @PatGet 👍
> 
> Once this is merged could you just give it a quick check with your application using a build from [our MyGet feed](https://www.myget.org/feed/aspnet-contrib/package/nuget/AspNet.Security.OAuth.Untappd) with the fix? If all is OK I'll publish a 5.0.3 release with the fix to NuGet.org tomorrow.

Will try it tomorrow morning.

**Comment by PatGet** [comment]
@martincostello Tested the MyGet, works fine.

**Pull request merged**
